### PR TITLE
Improve signature help precision

### PR DESCRIPTION
**Pull request by zth**
Better precision in signature help. You now don't have to start typing in an argument for it to highlight. We now look at commas to decide which argument you're currently intending to fill in.

Closes https://github.com/rescript-lang/rescript-vscode/issues/672

### CODE REVIEW

File: analysis/src/SignatureHelp.ml [hunk 1]
```diff
@@ -152,159 +152,206 @@ let docsForLabel typeExpr ~file ~package ~supportsMarkdownLinks =
   typeString :: typeDefinitions |> String.concat "\n"
 
 let signatureHelp ~path ~pos ~currentFile ~debug =
-  let posBeforeCursor = Pos.posBeforeCursor pos in
```
Comment (zth) [inline analysis/src/SignatureHelp.ml:155]: Absolutely terrible display of the diff in GitHub. I've added two optional unwrappings which messes with the indentation, which messes up the diff. Haven't really changed that much.

File: analysis/src/SignatureHelp.ml [hunk 2]
```diff
@@ -152,159 +152,206 @@ let docsForLabel typeExpr ~file ~package ~supportsMarkdownLinks =
...
+  | Some text -> (
+    match Pos.positionToOffset text pos with
+    | None -> None
+    | Some offset -> (
+      let posBeforeCursor = Pos.posBeforeCursor pos in
+      let offsetNoWhite = Utils.skipWhite text (offset - 1) in
+      let firstCharBeforeCursorNoWhite =
+        if offsetNoWhite < String.length text && offsetNoWhite >= 0 then
+          Some text.[offsetNoWhite]
+        else None
```
Comment (zth) [inline analysis/src/SignatureHelp.ml:167]: I added this.

File: analysis/src/SignatureHelp.ml [hunk 3]
```diff
@@ -152,159 +152,206 @@ let docsForLabel typeExpr ~file ~package ~supportsMarkdownLinks =
...
+                     if arg.exp.pexp_loc |> Loc.hasPos ~pos:posBeforeCursor then
+                       Some (Unlabelled currentUnlabelledArgCount)
+                     else (
+                       (* If this unlabelled arg doesn't have the cursor, record
+                          it as the last seen unlabelled arg before the
+                          cursor.*)
+                       if posBeforeCursor >= (arg.exp.pexp_loc |> Loc.start)
+                       then
+                         lastUnlabelledArgBeforeCursor :=
+                           currentUnlabelledArgCount;
```
Comment (zth) [inline analysis/src/SignatureHelp.ml:198]: This is new.

File: analysis/src/SignatureHelp.ml [hunk 4]
```diff
@@ -152,159 +152,206 @@ let docsForLabel typeExpr ~file ~package ~supportsMarkdownLinks =
...
+                 if firstCharBeforeCursorNoWhite = Some ',' then 1
+                   (* If we found no argument with the cursor, we might still be
+                      able to complete for an unlabelled argument, if the char
+                      before the cursor is ',', like: `someFn(123, <com>)`
+                      complete for argument 2, or: `someFn(123, <com>, true)`
+                      complete for argument 2 as well. Adding 1 here accounts
+                      for the comma telling us that the users intent is to fill
+                      in the next argument. *)
+                 else 0))
+          | v -> v
```
Comment (zth) [inline analysis/src/SignatureHelp.ml:237]: This is new.

**Review by cristianoc — APPROVED** [review]

**Pull request merged**